flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/LogicalPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+package org.apache.flink.runtime.executiongraph.failover.flip1;
+
+import org.apache.flink.runtime.jobgraph.topology.LogicalPipelinedRegion;
+import org.apache.flink.runtime.jobgraph.topology.LogicalResult;
+import org.apache.flink.runtime.jobgraph.topology.LogicalVertex;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.runtime.executiongraph.failover.flip1.PipelinedRegionComputeUtil.mergeRegions;
+import static org.apache.flink.runtime.executiongraph.failover.flip1.PipelinedRegionComputeUtil.mergeRegionsOnCycles;
+
+/** Utils for computing {@link LogicalPipelinedRegion}s. */
+public final class LogicalPipelinedRegionComputeUtil {
+    public static Set<Set<LogicalVertex>> computePipelinedRegions(
+            final Iterable<? extends LogicalVertex> topologicallySortedVertexes) {
+        final Map<LogicalVertex, Set<LogicalVertex>> vertexToRegion =
+                buildRawRegions(topologicallySortedVertexes);
+        return mergeRegionsOnCycles(
```

Comment:
Cycles should never happen in logical topology so I think the it is not needed.

Reply:
And then `PipelinedRegionComputeUtil#mergeRegions()` will only be needed by `SchedulingPipelinedRegionComputeUtil` and can be refactored into it.